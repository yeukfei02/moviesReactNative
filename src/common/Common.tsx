export const getRootUrl = () => {
  let ROOT_URL = '';

  if (process.env.NODE_ENV === 'development') {
    ROOT_URL = 'https://api.themoviedb.org/3';
  } else {
    ROOT_URL = 'https://api.themoviedb.org/3';
  }

  return ROOT_URL;
};
