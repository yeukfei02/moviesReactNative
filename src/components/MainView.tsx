import React from 'react';
import { StyleSheet, Text, View } from 'react-native';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

function MainView() {
  return (
    <View style={styles.container}>
      <Text>MainView</Text>
    </View>
  );
}

export default MainView;
