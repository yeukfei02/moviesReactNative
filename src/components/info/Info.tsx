import React, { useState, useEffect } from 'react';
import { StyleSheet, Text, ScrollView, View, Image, TouchableOpacity, Linking } from 'react-native';
import { MaterialIcons, FontAwesome } from '@expo/vector-icons';

import StackViewStatusBar from '../stackViewStatusBar/StackViewStatusBar';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f0f8ff',
  },
  viewContainer: {
    marginHorizontal: 30,
    marginVertical: 50,
  },
  titleText: {
    fontSize: 25,
    fontWeight: 'bold',
    textAlign: 'center',
    marginBottom: 10,
  },
  descriptionText: {
    fontSize: 20,
    textAlign: 'center',
    marginVertical: 10,
  },
  urlText: {
    fontSize: 20,
    textAlign: 'center',
    marginVertical: 10,
    color: 'blue',
    textDecorationLine: 'underline',
  },
});

function Info() {
  const handleTmdbLinkClick = () => {
    Linking.openURL(`https://developers.themoviedb.org/3/getting-started/introduction`);
  };

  const handleEmailClick = () => {
    Linking.openURL(`mailto:yeukfei02@gmail.com`);
  };

  const handleGithubClick = () => {
    Linking.openURL(`https://github.com/yeukfei02`);
  };

  return (
    <ScrollView style={styles.container} contentContainerStyle={{ flexGrow: 1 }}>
      <StackViewStatusBar backgroundColor="#3c5688" />

      <View style={styles.viewContainer}>
        <View style={{ alignItems: 'center' }}>
          <Image source={require('../../images/tmdb.png')} style={{ width: 250, height: 250, resizeMode: 'contain' }} />

          <View style={{ marginTop: 30 }}>
            <Text style={styles.titleText}>The Movie DB</Text>
            <Text style={styles.descriptionText}>All data are from The Movie DB developer</Text>
            <TouchableOpacity onPress={() => handleTmdbLinkClick()}>
              <Text style={styles.urlText}>https://developers.themoviedb.org/3/getting-started/introduction</Text>
            </TouchableOpacity>
          </View>

          <View style={{ marginVertical: 10 }}>
            <Text style={styles.descriptionText}>Developer: Donald Wu</Text>
            <View style={{ alignItems: 'center' }}>
              <View style={{ flexDirection: 'row' }}>
                <TouchableOpacity onPress={() => handleEmailClick()}>
                  <MaterialIcons name="email" size={50} color="black" />
                </TouchableOpacity>
                <TouchableOpacity style={{ marginHorizontal: 15 }} onPress={() => handleGithubClick()}>
                  <FontAwesome name="github" size={50} color="black" />
                </TouchableOpacity>
              </View>
            </View>
          </View>
        </View>
      </View>
    </ScrollView>
  );
}

export default Info;
