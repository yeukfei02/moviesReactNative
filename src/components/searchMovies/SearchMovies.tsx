import React, { useState, useEffect } from 'react';
import { StyleSheet, View, TextInput, ScrollView, FlatList, TouchableOpacity, Image, Text } from 'react-native';
import { Card } from 'react-native-paper';
import { useNavigation } from '@react-navigation/native';
import { FontAwesome } from '@expo/vector-icons';
import * as Network from 'expo-network';
import AsyncStorage from '@react-native-community/async-storage';
import axios from 'axios';
import _ from 'lodash';
import moment from 'moment';
import { TMDB_API_KEY } from 'react-native-dotenv';
import { getRootUrl } from '../../common/Common';

import SnackBar from '../snackBar/SnackBar';
import Divider from '../divider/Divider';
import StackViewStatusBar from '../stackViewStatusBar/StackViewStatusBar';

const ROOT_URL = getRootUrl();

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f0f8ff',
  },
  viewContainer: {
    marginHorizontal: 30,
    marginVertical: 35,
  },
  searchTextTextInput: {
    borderColor: 'black',
    borderRadius: 5,
    borderWidth: 1,
    padding: 15,
  },
  sortByRatingsButtonContainer: {
    marginVertical: 10,
    padding: 20,
    backgroundColor: '#b5c8ea',
    borderRadius: 5,
  },
  sortByRatingsButtonText: {
    color: 'black',
    textAlign: 'center',
    fontWeight: 'bold',
    textTransform: 'uppercase',
  },
  moviesItemCard: {
    paddingHorizontal: 25,
    paddingVertical: 20,
    marginVertical: 10,
  },
  moviesItemTitle: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  dividerStyle: {
    marginVertical: 5,
  },
  noDataContainer: {
    marginVertical: 10,
    padding: 20,
    backgroundColor: '#ff9800',
    borderRadius: 5,
  },
  noDataText: {
    fontSize: 15,
    color: 'black',
    textAlign: 'center',
    fontWeight: 'bold',
    textTransform: 'uppercase',
  },
  noNetworkContainer: {
    marginVertical: 10,
    padding: 20,
    backgroundColor: '#f44336',
    borderRadius: 5,
  },
  noNetworkText: {
    fontSize: 15,
    color: 'black',
    textAlign: 'center',
    fontWeight: 'bold',
    textTransform: 'uppercase',
  },
});

function MoviesItem(props: any) {
  const navigation = useNavigation();

  const handleMovieItemsClick = () => {
    navigation.navigate('Movie Details', {
      movieId: props.item.id,
    });
  };

  const renderRatingStar = (voteAverage: number) => {
    const ratingStar = [];

    for (let i = 0; i < voteAverage / 2; i++) {
      ratingStar.push(<FontAwesome key={i} name="star" size={18} color="#e5e500" style={{ marginHorizontal: 1 }} />);
    }

    if (ratingStar.length < 5) {
      ratingStar.push(<FontAwesome name="star-o" size={18} color="black" style={{ marginHorizontal: 1 }} />);
    }

    return ratingStar;
  };

  return (
    <Card style={styles.moviesItemCard}>
      <TouchableOpacity onPress={() => handleMovieItemsClick()}>
        <View style={{ flexDirection: 'row' }}>
          <Image
            source={{ uri: `https://image.tmdb.org/t/p/w500${props.item.poster_path}` }}
            style={{ width: 80, height: 120, resizeMode: 'stretch' }}
          />
          <View style={{ flexDirection: 'column', justifyContent: 'center', marginHorizontal: 10 }}>
            <Text style={styles.moviesItemTitle}>{props.item.title}</Text>
            <View style={{ flexDirection: 'row', marginVertical: 10 }}>
              {renderRatingStar(props.item.vote_average)}
            </View>
          </View>
        </View>
      </TouchableOpacity>
    </Card>
  );
}

function SearchMovies(props: any) {
  const [hasNetwork, setHasNetwork] = useState(false);

  const [searchText, setSearchText] = useState('');
  const [moviesListData, setMoviesListData] = useState<any[]>([]);
  const [page, setPage] = useState(1);

  const [snackBarStatus, setSnackBarStatus] = useState(false);
  const [snackBarType, setSnackBarType] = useState('');
  const [snackBarMessage, setSnackBarMessage] = useState('');

  useEffect(() => {
    checkNetworkStatus();
    getAsyncStorageData();
  }, []);

  useEffect(() => {
    if (hasNetwork) {
      if (searchText && searchText.length > 3) {
        getMoviesListData(searchText);
      } else {
        setMoviesListData([]);
      }
    }
  }, [hasNetwork, searchText]);

  const checkNetworkStatus = async () => {
    const networkStatus = await Network.getNetworkStateAsync();
    if (networkStatus.isConnected && networkStatus.isInternetReachable) {
      setHasNetwork(true);
    }
  };

  const getAsyncStorageData = async () => {
    try {
      const searchTextFromAsyncStorage = await AsyncStorage.getItem('@searchText');
      if (searchTextFromAsyncStorage) {
        setSearchText(searchTextFromAsyncStorage);
      }
    } catch (e) {
      console.log('error = ', e.message);
    }
  };

  const storeAsyncStorageData = async (key: string, value: any) => {
    try {
      await AsyncStorage.setItem(key, value);
    } catch (e) {
      console.log('error = ', e.message);
    }
  };

  const getMoviesListData = async (searchText: string) => {
    const response = await axios.get(`${ROOT_URL}/movie/popular`, {
      params: {
        api_key: TMDB_API_KEY,
        language: 'en-US',
        page: page,
      },
    });
    const responseData = response.data;
    console.log('responseData = ', responseData);

    if (responseData.results) {
      const currentYear = moment();
      const lastYear = moment().subtract(1, 'years');
      const filteredResults = responseData.results.filter((item: any, i: number) => {
        const releaseYear = moment(item.release_date);
        if (item.title.includes(searchText) && releaseYear.isBetween(lastYear, currentYear)) {
          return item;
        }
      });
      console.log('filteredResults = ', filteredResults);
      setMoviesListData(filteredResults);
    }
  };

  const handleSearchTextChange = (searchText: string) => {
    storeAsyncStorageData('@searchText', searchText);
    setSearchText(searchText);
  };

  const renderMoviesItem = (props: any) => {
    return <MoviesItem item={props.item} />;
  };

  const renderResultDiv = (hasNetwork: boolean) => {
    let resultDiv = (
      <View style={styles.noNetworkContainer}>
        <Text style={styles.noNetworkText}>There are no network</Text>
      </View>
    );

    if (hasNetwork) {
      resultDiv = (
        <View>
          {renderSortByRatingsButton(moviesListData)}
          {renderMoviesList(moviesListData)}
        </View>
      );
    }

    return resultDiv;
  };

  const renderSortByRatingsButton = (moviesListData: any[]) => {
    let sortByRatingsButton = null;

    if (!_.isEmpty(moviesListData)) {
      sortByRatingsButton = (
        <TouchableOpacity onPress={() => handleSortByRatings()}>
          <View style={styles.sortByRatingsButtonContainer}>
            <Text style={styles.sortByRatingsButtonText}>Sort by ratings</Text>
          </View>
        </TouchableOpacity>
      );
    }

    return sortByRatingsButton;
  };

  const renderMoviesList = (moviesListData: any[]) => {
    let moviesList = (
      <View style={styles.noDataContainer}>
        <Text style={styles.noDataText}>There are no data</Text>
      </View>
    );

    if (!_.isEmpty(moviesListData)) {
      moviesList = (
        <FlatList
          data={moviesListData}
          renderItem={renderMoviesItem}
          keyExtractor={(item, index) => index.toString()}
        />
      );
    }

    return moviesList;
  };

  const handleSortByRatings = () => {
    if (!_.isEmpty(moviesListData)) {
      const sortedMoviesListData = _.orderBy(moviesListData, ['vote_average'], ['desc']);
      setMoviesListData(sortedMoviesListData);
    }
  };

  const handleSnackBarDismiss = () => {
    if (snackBarStatus) {
      setSnackBarStatus(false);
    }
  };

  return (
    <ScrollView style={styles.container} contentContainerStyle={{ flexGrow: 1 }}>
      <StackViewStatusBar backgroundColor="#3c5688" />

      <View style={styles.viewContainer}>
        <TextInput
          style={styles.searchTextTextInput}
          placeholder="Enter some text..."
          placeholderTextColor="black"
          value={searchText}
          onChangeText={(text: string) => handleSearchTextChange(text)}
        />

        <Divider style={styles.dividerStyle} />

        {renderResultDiv(hasNetwork)}
      </View>

      <SnackBar
        snackBarStatus={snackBarStatus}
        snackBarType={snackBarType}
        handleSnackBarDismiss={() => handleSnackBarDismiss()}
        snackBarMessage={snackBarMessage}
      />
    </ScrollView>
  );
}

export default SearchMovies;
