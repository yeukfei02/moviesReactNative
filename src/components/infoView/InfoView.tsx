import React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';

import Info from '../info/Info';

const Stack = createStackNavigator();

function InfoView(): JSX.Element {
  return (
    <NavigationContainer independent={true}>
      <Stack.Navigator
        screenOptions={{
          headerStyle: {
            backgroundColor: '#b5c8ea',
          },
          headerTitleStyle: {
            color: 'black',
          },
        }}
      >
        <Stack.Screen name="Info" component={Info} />
      </Stack.Navigator>
    </NavigationContainer>
  );
}

export default InfoView;
