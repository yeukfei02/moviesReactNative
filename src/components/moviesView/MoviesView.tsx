import React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';

import SearchMovies from '../searchMovies/SearchMovies';
import MoviesDetails from '../moviesDetails/MoviesDetails';

const Stack = createStackNavigator();

function MoviesView() {
  return (
    <NavigationContainer independent={true}>
      <Stack.Navigator
        screenOptions={{
          headerStyle: {
            backgroundColor: '#b5c8ea',
          },
          headerTitleStyle: {
            color: 'black',
          },
        }}
      >
        <Stack.Screen name="Movies" component={SearchMovies} />
        <Stack.Screen name="Movies Details" component={MoviesDetails} />
      </Stack.Navigator>
    </NavigationContainer>
  );
}

export default MoviesView;
