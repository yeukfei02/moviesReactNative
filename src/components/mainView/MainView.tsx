import React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { Feather, MaterialCommunityIcons } from '@expo/vector-icons';

import MoviesView from '../moviesView/MoviesView';
import InfoView from '../infoView/InfoView';

const Tab = createBottomTabNavigator();

function MainView(): JSX.Element {
  return (
    <NavigationContainer independent={true}>
      <Tab.Navigator
        screenOptions={({ route }) => ({
          tabBarIcon: ({ focused, color, size }) => {
            let icon = null;

            switch (route.name) {
              case 'Movies':
                icon = focused ? (
                  <MaterialCommunityIcons name="movie" size={size} color={color} />
                ) : (
                  <MaterialCommunityIcons name="movie" size={size} color={color} />
                );
                break;
              case 'Info':
                icon = focused ? (
                  <Feather name="info" size={size} color={color} />
                ) : (
                  <Feather name="info" size={size} color={color} />
                );
                break;
            }

            return icon;
          },
        })}
        tabBarOptions={{
          activeTintColor: '#3c5688',
          inactiveTintColor: 'gray',
        }}
      >
        <Tab.Screen name="Movies" component={MoviesView} />
        <Tab.Screen name="Info" component={InfoView} />
      </Tab.Navigator>
    </NavigationContainer>
  );
}

export default MainView;
