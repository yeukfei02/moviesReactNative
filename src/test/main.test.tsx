import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount } from 'enzyme';
import mockAsyncStorage from '@react-native-community/async-storage/jest/async-storage-mock';
import axios from 'axios';
import { TMDB_API_KEY } from 'react-native-dotenv';
import { getRootUrl } from '../common/Common';

import MainView from '../components/mainView/MainView';

import MoviesView from '../components/moviesView/MoviesView';
import InfoView from '../components/infoView/InfoView';

import SearchMovies from '../components/searchMovies/SearchMovies';
// import MovieDetails from '../components/movieDetails/MovieDetails';

import Info from '../components/info/Info';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('@react-native-community/async-storage', () => mockAsyncStorage);

const ROOT_URL = getRootUrl();
let movieId = '';

describe('main.test', () => {
  describe('render test', () => {
    it('MainView', () => {
      const wrapper = shallow(<MainView />);
      expect(wrapper).toMatchSnapshot();
    });

    it('MoviesView', () => {
      const wrapper = shallow(<MoviesView />);
      expect(wrapper).toMatchSnapshot();
    });

    it('InfoView', () => {
      const wrapper = shallow(<InfoView />);
      expect(wrapper).toMatchSnapshot();
    });

    it('SearchMovies', () => {
      const wrapper = shallow(<SearchMovies />);
      expect(wrapper).toMatchSnapshot();
    });

    // it('MovieDetails', () => {
    //   const wrapper = shallow(<MovieDetails />);
    //   expect(wrapper).toMatchSnapshot();
    // });

    it('Info', () => {
      const wrapper = shallow(<Info />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('api request test', () => {
    it('search movies request test', async () => {
      const response = await axios.get(`${ROOT_URL}/search/movie`, {
        params: {
          api_key: TMDB_API_KEY,
          query: 'alad',
          language: 'en-US',
          page: 1,
        },
      });
      const responseData = response.data;
      // console.log('responseData = ', responseData);

      expect(responseData).toBeDefined();
      expect(responseData.results).toBeDefined();

      movieId = responseData.results[0].id;

      if (responseData) {
        responseData.results.forEach((item: any, i: number) => {
          expect(item.id).toBeDefined();
          expect(typeof item.id).toBe('number');

          expect(item.title).toBeDefined();
          expect(typeof item.title).toBe('string');

          expect(item.poster_path).toBeDefined();
          // expect(typeof item.poster_path).toBe('string');

          expect(item.vote_average).toBeDefined();
          expect(typeof item.vote_average).toBe('number');
        });
      }
    });

    it('movies details request test', async () => {
      const response = await axios.get(`${ROOT_URL}/movie/${movieId}`, {
        params: {
          api_key: TMDB_API_KEY,
          language: 'en-US',
        },
      });
      const responseData = response.data;
      // console.log('responseData = ', responseData);

      expect(responseData).toBeDefined();

      if (responseData) {
        expect(responseData.poster_path).toBeDefined();
        expect(typeof responseData.poster_path).toBe('string');

        expect(responseData.title).toBeDefined();
        expect(typeof responseData.title).toBe('string');

        expect(responseData.overview).toBeDefined();
        expect(typeof responseData.overview).toBe('string');

        expect(responseData.release_date).toBeDefined();
        expect(typeof responseData.release_date).toBe('string');

        expect(responseData.vote_average).toBeDefined();
        expect(typeof responseData.vote_average).toBe('number');

        expect(responseData.vote_count).toBeDefined();
        expect(typeof responseData.vote_count).toBe('number');

        expect(responseData.popularity).toBeDefined();
        expect(typeof responseData.popularity).toBe('number');
      }
    });
  });
});
