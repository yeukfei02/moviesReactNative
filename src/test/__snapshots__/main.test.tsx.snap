// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main.test render test Info 1`] = `ShallowWrapper {}`;

exports[`main.test render test InfoView 1`] = `ShallowWrapper {}`;

exports[`main.test render test MainView 1`] = `ShallowWrapper {}`;

exports[`main.test render test MoviesView 1`] = `ShallowWrapper {}`;

exports[`main.test render test SearchMovies 1`] = `ShallowWrapper {}`;
